spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,395 @@
+ * existing Truncate partition transform, e.g. {@code SELECT system.truncate(1, 'abc')} will return
+ * the String 'a'.
+ *
+ * <p>Note that for performance reasons, the given input width is not validated in the
+ * implementations used in code-gen. The width must remain non-negative to give meaningful results.
+ */
+public class TruncateFunction implements UnboundFunction {
+  private static final List<DataType> truncateableAtomicTypes =
+      ImmutableList.of(
+          DataTypes.ByteType,
+          DataTypes.ShortType,
+          DataTypes.IntegerType,
+          DataTypes.LongType,
+          DataTypes.StringType,
+          DataTypes.BinaryType);
+
+  private static void validateTruncationFieldType(DataType dt) {
+    if (truncateableAtomicTypes.stream().noneMatch(type -> type.sameType(dt))
+        && !(dt instanceof DecimalType)) {
+      String expectedTypes =
+          "[ByteType, ShortType, IntegerType, LongType, StringType, BinaryType, DecimalType]";
+      throw new UnsupportedOperationException(
+          String.format(
+              "Invalid input type to truncate. Expected one of %s, but found %s",
+              expectedTypes, dt));
+    }
+  }
+
+  private static void validateTruncationWidthType(DataType widthType) {
+    if (!DataTypes.IntegerType.sameType(widthType)
+        && !DataTypes.ShortType.sameType(widthType)
+        && !DataTypes.ByteType.sameType(widthType)) {
+      throw new UnsupportedOperationException(
+          "Expected truncation width to be one of [ByteType, ShortType, IntegerType], but found "
+              + widthType);
+    }
+  }
+
+  @Override
+  public BoundFunction bind(StructType inputType) {
+    if (inputType.fields().length != 2) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Invalid input type. Expected 2 fields but found %s", inputType.fields().length));
+    }
+
+    StructField widthField = inputType.apply(0);
+    StructField toTruncateField = inputType.apply(1);
+
+    validateTruncationFieldType(toTruncateField.dataType());
+    validateTruncationWidthType(widthField.dataType());
+
+    DataType toTruncateDataType = toTruncateField.dataType();
+    if (toTruncateDataType instanceof ByteType) {
+      return new TruncateTinyInt();
+    } else if (toTruncateDataType instanceof ShortType) {
+      return new TruncateSmallInt();
+    } else if (toTruncateDataType instanceof IntegerType) {
+      return new TruncateInt();
+    } else if (toTruncateDataType instanceof LongType) {
+      return new TruncateBigInt();
+    } else if (toTruncateDataType instanceof DecimalType) {
+      return new TruncateDecimal(
+          ((DecimalType) toTruncateDataType).precision(),
+          ((DecimalType) toTruncateDataType).scale());
+    } else if (toTruncateDataType instanceof StringType
+        || toTruncateDataType instanceof VarcharType
+        || toTruncateDataType instanceof CharType) {
```

Comment:
Varchar and Char aren't allowed by `validateTruncationFieldType`. Do these `instanceof` checks work?

Reply:
Yes these `instanceof` checks work the same way that that `instanceof DecimalType` works (because `VarcharType` and `CharType` both have a size / length field so they cannot be compared directly).

I can update `validateTruncationFieldType`. However, I think that `validateTruncationFieldType` is somewhat unnecesssary given that this whole block basically does the same thing. Let me try moving it around a bit.

Reply:
Yeah, I don't think that method is necessary. Let's remove it.

Reply:
Removed the method.

I'm going to remove `varchartype` and `chartype` as Spark prints a warning saying that it casts these two aliases for `UTF8String` to string already and that neither type can be used for functions etc.

Reply:
Yep. Removed the check for `VarcharType` and `CharType` as they're not necessary.

There are unit tests to verify this behavior and it also matches the Spark docs.